### PR TITLE
Use SmartyStreets geocoding API for addresses

### PR DESCRIPTION
**Pull request by kairstenfay**
EDIT: see latest commit message for updated info.

vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
These changes swapped Google's geocoding API for SmartyStreet's API.
The HTTP GET request requires an authentication token and ID which are
not tracked through git and need to be added in a separate file,
`environment.py`. An example of a valid `environment.py` configuration
is viewable at `example.environment.py`. These authentication strings
are currently imported as global variables in
`address_to_census_tract.py`.

The program currently works by passing SmartyStreets a free text
address via the `street` parameter. There are a couple of assumptions
made about addresses in order for this program to work. Namely,
    1. A street address contains at least one letter, [a-z] (case
    insensitive).
    2. A street address comes before a zip code in an address.

We do not assume that an address is comma-separated.

If an address lookup fails on the first attempt (receives a status code
of 200 but returns no JSON object), then a second attempt is made to
lookup the address. In the second attempt, the program
    1. Splits the address (string) on any commas, converting it into a
    list.
    2. Removes the first index of the list if it contains only numbers.
    For example, in the original address `100, 123 Main St.`, `100`
    would be removed from the list.
    3. Finds and removes a partial string matching the pattern
    `#.*(?\s)` (a pound sign followed by anything with an optional
    space). In the original address `123 Main St. #100`, `#100` would be
    removed.

Removing these unit or suite numbers was essential in getting the
SmartyStreets API to find matches at certain addresses in the
`testset.json` file.

### CODE REVIEW

**Comment by kairstenfay** [comment]
[Standalone address de-identification program](https://trello.com/c/SQW5WRKs/20-standalone-address-de-identification-program)

File: address_to_census_tract.py [hunk 1]
```diff
@@ -7,81 +7,194 @@
...
-    geocoder
     shapely
 """
-import geocoder
 import json
 from shapely.geometry import shape, Point
-from sys import stderr
-
+from sys import stderr, argv
+from environment import AUTH_ID, AUTH_TOKEN  # Global variables
```
Comment (tsibley) [inline address_to_census_tract.py]: Let's make these environment variables instead of Python globals? For example, `os.environ.get("SMARTYSTREETS_AUTH_ID")` instead of importing `AUTH_ID` from an `environment.py`.

File: address_to_census_tract.py [hunk 2]
```diff
@@ -7,81 +7,194 @@
...
+import re
+import requests
+import pandas as pd
```
Comment (tsibley) [inline address_to_census_tract.py:24]: Do you know about [SmartyStreet's Python SDK](https://smartystreets.com/docs/sdk/python)? We may want to use that instead of rolling our own API client.

File: address_to_census_tract.py [hunk 3]
```diff
@@ -7,81 +7,194 @@
...
-    """Convert an address string to a list of latitude, longitude coordinates.
-
-    Currently uses Google's geocoder API for geocoding, but this could be
-    replaced with other geocoder implementations.
+def process_csv(file_path):
     """
-    return geocoder.google(address).latlng
-
+    This is going to prove difficult if addresses are separated by commas.
+    Excel may be the only way to go unless all addresses are wrapped in quotes.
```
Comment (tsibley) [inline address_to_census_tract.py]: The CSV isn't valid if a field value contains unescaped/unquoted commas. I don't think we have to worry about this.

File: address_to_census_tract.py [hunk 4]
```diff
@@ -7,81 +7,194 @@
...
+
+def response_to_standard_address(response: object) -> str:
+    """
+    Returns a compiled and standard address from the given HTTP response object.
+    """
+    standard_address = response['delivery_line_1']
 
-    with open(geojson_filename) as file:
-        geojson = json.load(file)
+    if 'delivery_line_2' in response.keys():
```
Comment (tsibley) [inline address_to_census_tract.py]: FWIW, the `in` operator works directly against dictionaries, so the `.keys()` is unnecessary.

File: address_to_census_tract.py [hunk 5]
```diff
@@ -7,81 +7,194 @@
...
 
-def latlng_to_polygon(latlng, polygons):
+def latlng_to_polygon(latlng):
     """
-    Find the first polygon in *polygons* which contains the *latlng* and return
-    the polygon, else None.
+    Find the first polygon in *polygons* (loaded from file) which contains the 
+    *latlng* and return the polygon, else None.
     """
+    polygons = load_geojson("data/geojsons/Washington_2016.geojson")
```
Comment (tsibley) [inline address_to_census_tract.py]: Moving the loading of the geojson into the lat/lng to polygon function means that the geojson is loaded again for every single lookup. That's going to slow things down a bit, so we should lift it up.

File: address_to_census_tract.py [hunk 6]
```diff
@@ -7,81 +7,194 @@
...
+    numbers often cause errors in geocoding.  
+
+    Assumptions:
+    1. A street address contains at least one letter, [a-z] (case insensitive).
+    2. A street address comes before a city, state, and zip code in an address.
+    3. An address won't have more than one comma-separated string ahead of a 
+    true street address.
+
+    The following method should work even if the given address has no commas
+    separating the address components.
```
Comment (tsibley) [inline address_to_census_tract.py]: I'm not comfortable with the accuracy of trying to parse and reformat addresses ourselves. That's part of reason we're using a third-party API. :-) I wonder why SmartyStreets fails on these? Could it be what we're asking for or the way we're doing a request rather than some fundamental inability? Can we consult with SmartyStreets documentation/support here for a better way to handle the problematic addresses?

What other solutions did you try?

File: address_to_census_tract.py [hunk 7]
```diff
@@ -7,81 +7,194 @@
...
+def build_request_url(address: str) -> str:
+    """
+    Given a dict containing an address, builds a request URL for the
+    SmartyStreets geocoding API
+    """
+    NUM_CANDIDATES = 1
+
+    url = 'https://us-street.api.smartystreets.com/street-address?'
+    url += f'auth-id={AUTH_ID}&auth-token={AUTH_TOKEN}&street='
+    url += f'{address}&candidates={NUM_CANDIDATES}'
```
Comment (tsibley) [inline address_to_census_tract.py]: The interpolated values aren't URL-encoded, which means they'll break if the address (or id/token) contains a reserved character, such as in `4th & Walnut`. Note that `requests` supports ways to pass a dict to generate query or body params, but also that this function won't be necessary if we use the SmartyStreets SDK.

File: address_to_census_tract.py [hunk 8]
```diff
@@ -7,81 +7,194 @@
...
-    return geocoder.google(address).latlng
-
+    This is going to prove difficult if addresses are separated by commas.
+    Excel may be the only way to go unless all addresses are wrapped in quotes.
+    """
+    df = pd.read_csv(file_path)
+    df['response'] = df['address'].apply(lookup_address)
+    df['standard_address'] = df['response'].apply(response_to_standard_address)
+    df['latlng'] = df['response'].apply(response_to_latlng)
+    df['final_tract'] = df['latlng'].apply(latlng_to_polygon)
```
Comment (tsibley) [inline address_to_census_tract.py]: Latitude and longitude should end up as separate columns in the CSV.

File: address_to_census_tract.py [hunk 9]
```diff
@@ -7,81 +7,194 @@
...
+    df.to_csv('data/test/out.csv') 
+
+def lookup_address(address: str) -> dict:
+    """
+    Given an address, returns a dict containing geospatial information and 
+    metadata from the SmartyStreet's geocoding API.
 
-def load_geojson(geojson_filename):
-    """Read GeoJSON file and return a list of features converted to shapes."""
+    To minimize cost, a response for each address is only looked up once.
```
Comment (tsibley) [inline address_to_census_tract.py]: This claim in the docstring makes me think that the function is employing caching/memoization. The comment is more applicable to the _callers_ of this function, which are what control how many times an address is looked up.

**Review by tsibley — CHANGES_REQUESTED** [review]

File: address_to_census_tract.py (hunk 1 above)
Comment (kairstenfay) [reply address_to_census_tract.py]: Done and loaded upon conda environment activation, per https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#macos-and-linux

File: address_to_census_tract.py (hunk 2 above)
Comment (kairstenfay) [reply address_to_census_tract.py:24]: Now using `smartystreets_python_sdk`!

File: address_to_census_tract.py (hunk 7 above)
Comment (kairstenfay) [reply address_to_census_tract.py]: Resolving in favor of using SmartyStreets SDK.

File: address_to_census_tract.py (hunk 6 above)
Comment (kairstenfay) [reply address_to_census_tract.py]: Your concerns make sense.

Incidentally, switching to the smartstreets-python-sdk library proved more robust for addresses with the Suite or Unit # issue. 

I'm still looking into fixing the lookup for the other problematic address (with unit number out front, plus a comma).

Comment (kairstenfay) [reply address_to_census_tract.py]: I switched to using SmartyStreets' US Extract API for addresses where the normal lookup fails.

**Comment by kairstenfay** [comment]
Please see the new changes I've implemented per the previous code review.

File: address_to_census_tract.py [hunk 10]
```diff
@@ -7,81 +7,167 @@
...
-    with open("data/test/testset.json", encoding = "UTF-8") as file:
+    if argv[1].endswith('.json'):
+        process_json(argv[1])  # hint: "data/test/testset.json"
+    elif argv[1].endswith('.csv'):
+        process_csv(argv[1])  # hint: "data/test/testset.csv"
+    else:
+        raise ValueError(f"Unknown file extension for file named «{argv[1]}». \
+Please choose from one of the following file extensions: \n \
+    * .csv \n \
+    * .json")
```
Comment (tsibley) [inline address_to_census_tract.py]: FYI only, not necessary to change:

It can sometimes be nicer to write multi-line strings using Python's triple-quote syntax and [the `dedent` function from the textwrap module](https://docs.python.org/3/library/textwrap.html#textwrap.dedent). For example:

```python
        raise ValueError(dedent(f"""
            Unknown file extension for file named «{argv[1]}».
            Please choose from one of the following file extensions:
                * .csv
                * .json
        """))
```

File: address_to_census_tract.py [hunk 11]
```diff
@@ -7,81 +7,167 @@
...
         address = record["address"]
 
+        standard_address = None
+        latlng = None
         tract = None
-        latlng = address_to_latlng(address)
+        response = lookup_address(address)
 
-        if latlng:
+        if response and response['lat']:  # assume 'lng' will be present too
```
Comment (tsibley) [inline address_to_census_tract.py]: Why assume and not test?

File: address_to_census_tract.py [hunk 12]
```diff
@@ -7,81 +7,167 @@
...
+    except exceptions.SmartyException as err:
+        print(err)
+        return
+
+    result = lookup.result
+    if not result:  # Invalid address
+        result = extract_address(credentials, address)
+
+    first_candidate = result[0]
+    return {"standard_address": first_candidate.delivery_line_1 + ' ' +
```
Comment (tsibley) [inline address_to_census_tract.py]: What about `delivery_line_2`?

File: address_to_census_tract.py [hunk 13]
```diff
@@ -7,81 +7,167 @@
...
+    lookup = Lookup()
+    lookup.street = address
+    lookup.candidates = 1
+    lookup.match = "Invalid"  # Most permissive
+    
+    try:
+        client.send_lookup(lookup)
+    except exceptions.SmartyException as err:
+        print(err)
+        return
```
Comment (tsibley) [inline address_to_census_tract.py:129]: Sending the error to stdout means it'll be mixed in with the CSV or JSON output.

**Review by tsibley — CHANGES_REQUESTED** [review]

File: address_to_census_tract.py (hunk 10 above)
Comment (kairstenfay) [reply address_to_census_tract.py]: Good to know! I'll use that from now on.

File: address_to_census_tract.py (hunk 11 above)
Comment (kairstenfay) [reply address_to_census_tract.py]: Blind faith? I'll add that second test.

File: address_to_census_tract.py (hunk 12 above)
Comment (kairstenfay) [reply address_to_census_tract.py]: I wasn't sure if we needed that level of granularity, but I can easily add it back in.

File: address_to_census_tract.py (hunk 13 above)
Comment (kairstenfay) [reply address_to_census_tract.py:129]: Is this better handled by
`LOG.exception(err)` ?

File: .gitignore [hunk 14]
```diff
@@ -7,3 +7,5 @@ environment*
 data/geojsons/*
 
 *.DS_Store
+
+/__pycache__
```
Comment (tsibley) [inline .gitignore]: This change should really be in a separate commit from the README updates. The leading slash should also be removed so that `__pycache__` directories anywhere are ignored, not just the top-level one. Cached compiled Python modules should never be committed to version control.

File: README.md [hunk 15]
```diff
@@ -2,57 +2,61 @@
 
 `address_to_census_tract.py` prototypes address to Census tract conversion.  It uses a two-step process:
 
-1. Geocode a given single-line address to a (latitude, longitude) pair using a
-   remote (forward) geocoding service.
+1. Geocode a given single-line address to a (latitude, longitude) pair using a remote (forward) 
+   geocoding service.
```
Comment (tsibley) [inline README.md]: A lot of the changes to this README are re-wrapping lines from ~80 chars to a longer length. Doing so while making content changes makes the diff noisy and hard to read.

The actual content changes should also be part of the first commit on the branch.

File: address_to_census_tract.py [hunk 16]
```diff
@@ -4,84 +4,194 @@
...
+
+LOG = logging.getLogger(__name__)
 
 
 def main():
-    with open("data/test/testset.json", encoding = "UTF-8") as file:
+    if argv[1].endswith('.json'):
+        process_json(argv[1])
+    elif argv[1].endswith('.csv'):
+        process_csv(argv[1])
```
Comment (tsibley) [inline address_to_census_tract.py:38]: The number of args should be checked before accessing `argv[1]`. Perhaps we should switch to using argparse now.

File: address_to_census_tract.py [hunk 17]
```diff
@@ -4,84 +4,194 @@
...
         }
 
         print(json.dumps({ **record, "result": result }))
 
+def process_csv(file_path: str):
+    """
+    Given a *file_path* to a CSV, processes the `address` column and adds 
+    extra columns for a standardized address, latitude and longitude
+    coordinates, and census tract. If a given address is invalid, these new 
+    columns are left blank. Saves the new table at `data/test/out.csv`.
```
Comment (tsibley) [inline address_to_census_tract.py]: How about outputting the CSV to stdout instead of a hardcoded test file name? (Happy to explain why I think this preferred.)

**Review by tsibley — CHANGES_REQUESTED** [review]

File: address_to_census_tract.py (hunk 17 above)
Comment (kairstenfay) [reply address_to_census_tract.py]: I understand why a hardcoded test file name is not preferred, but why to stdout? Would you rather save the data via
`python address_to_census_tract.py {input_file} > {output_file}`
?

Comment (tsibley) [reply address_to_census_tract.py]: The convention of Unix command-line tools is to take input on stdin or named files and send output to stdout or named files.

I suggested stdout because the JSON output is already sent to stdout, and shell redirection is an easy way to save the output. Adding an `--output / -o` option would also be good to name the output file explicitly, but that doesn't have to happen in this PR unless you want to.

Comment (kairstenfay) [reply address_to_census_tract.py]: I like that idea. I'll switch to stdout for now and incorporate output options into the next PR.

**Comment by kairstenfay** [comment]
I fixed the diffs so they're much more intelligible! Also, I'm now dumping the CSV to stdout and checking for the length of arguments before processing the input file.
